### PR TITLE
Add support variation risk balancing

### PR DESCRIPTION
**Pull request by charstal**
1. I add a yaml deploy file
2. finished the support for second plugin simply

testing result
image
image

I havn't finished the testing file

```json
{"timestamp":1609843956,"window":{"duration":"15m","start":1609843056,"end":1609843956},"source":"","data":{"NodeMetricsMap":{"k8s-1m-2n-latest-worker":{"metrics":[{"name":"host.cpu.utilisation","type":"CPU","rollup":"AVG","value":0.003616666666666667},{"name":"host.cpu.utilisation","type":"CPU","rollup":"STD","value":0.00000004345238095238095},{"name":"host.memory.utilisation","type":"Memory","rollup":"AVG","value":0.01586203949225173},{"name":"host.memory.utilisation","type":"Memory","rollup":"STD","value":0.0000000009411688868542183}],"tags":{},"metadata":{"dataCenter":""}},"k8s-1m-2n-latest-worker2":{"metrics":[{"name":"host.cpu.utilisation","type":"CPU","rollup":"AVG","value":0.004158333333333333},{"name":"host.cpu.utilisation","type":"CPU","rollup":"STD","value":0.00000787157738095238},{"name":"host.memory.utilisation","type":"Memory","rollup":"AVG","value":0.01659691465281302},{"name":"host.memory.utilisation","type":"Memory","rollup":"STD","value":0.00000005600158888731594}],"tags":{},"metadata":{"dataCenter":""}},"k8s-1m-2n-latest-control-plane":{"metrics":[{"name":"host.cpu.utilisation","type":"CPU","rollup":"AVG","value":0.022574999999999998},{"name":"host.cpu.utilisation","type":"CPU","rollup":"STD","value":0.000001104464285714288},{"name":"host.memory.utilisation","type":"Memory","rollup":"AVG","value":0.05292091717184762},{"name":"host.memory.utilisation","type":"Memory","rollup":"STD","value":0.000000025921568293925273}],"tags":{},"metadata":{"dataCenter":""}}}}}
```

### CODE REVIEW

**Comment by ridv** [comment]
@charstal some folks working on this are on vacation so you may not get feedback quick but it'll come 😄

File: pkg/metricsprovider/signalfx.go [hunk 1]
```diff
@@ -64,13 +64,13 @@ func NewSignalFxClient() (watcher.FetcherClient, error) {
...
 
-func (s signalFxClient) FetchHostMetrics(host string, window *watcher.Window) ([]watcher.Metric, error) {
+func (s signalFxClient) FetchHostMetrics(host string) ([]watcher.Metric, error) {
 	log.Debugf("fetching metrics for host %v", host)
 	var metrics []watcher.Metric
 	hostQuery := signalFxHostFilter + host + signalFxHostNameSuffix
 
 	for _, metric := range []string{cpuUtilizationMetric, memoryUtilizationMetric} {
-		uri, err := buildMetricURL(hostQuery, metric, window)
+		uri, err := buildMetricURL(hostQuery, metric)
```
Comment (wangchen615) [inline pkg/metricsprovider/signalfx.go:73]: Why you remove the `window` variable?

File: pkg/watcher/watcher.go [hunk 2]
```diff
@@ -33,26 +33,31 @@ import (
...
-	mutex         sync.RWMutex // For thread safe access to cache
-	fifteenMinute []WatcherMetrics
-	tenMinute     []WatcherMetrics
-	fiveMinute    []WatcherMetrics
-	cacheSize     int
-	client        FetcherClient
-	isStarted     bool // Indicates if the Watcher is started by calling StartWatching()
-	shutdown      chan os.Signal
+	mutex       sync.RWMutex // For thread safe access to cache
+	minuteQueue []WatcherMetrics
```
Comment (wangchen615) [inline pkg/watcher/watcher.go:56]: I believe the rolling window should be configurable, at least giving several options.

File: pkg/watcher/watcher.go [hunk 3]
```diff
@@ -243,11 +347,59 @@ func (w *Watcher) deepCopyWatcherMetrics(src *WatcherMetrics) *WatcherMetrics {
 	}
 }
 
-// HTTP Handler for baseUrl endpoint
-func (w *Watcher) handler(resp http.ResponseWriter, r *http.Request) {
+// HTTP Handler for loadVariationRiskBalancing endpoint
```
Comment (wangchen615) [inline pkg/watcher/watcher.go:350]: We should discuss if we use a generic handler with different params for all plugins or use different handlers.

**Review by wangchen615 — COMMENTED** [review]
@charstal, thanks so much for the contribution. I went through the code and left my comments. There are at least two issues we need to discuss with @zorro786. 
1. Whether we remove window options completely?
2. Whether we use different handlers or use the same handler but limit the data scope by parameters?

File: pkg/watcher/watcher.go (hunk 3 above)
Comment (charstal) [reply pkg/watcher/watcher.go:350]: yes I do agree with you, but how to set and set for what

File: pkg/watcher/watcher.go (hunk 2 above)
Comment (charstal) [reply pkg/watcher/watcher.go:56]: ok, it should be configurable, let's add some envs to loadwatcher?

File: pkg/metricsprovider/signalfx.go (hunk 1 above)
Comment (charstal) [reply pkg/metricsprovider/signalfx.go:73]: before your email, I think it's useless, it means setting for Prometheus?

**Comment by zorro786** [comment]
@charstal Thank you for your contribution, it is not clear to me the need for this PR. Please open an issue first and we can discuss there. For your information, @wangchen615 is working to add support for "LoadVariationRiskBalancing" plugin, this will take care of memory/std/avg etc. I am currently working on a hands on tutorial which includes deployment spec for Trimaran Scheduler. I am closing this PR for now; if you would like to contribute for anything please open an issue and we can take it from there.

**Pull request closed without merging**